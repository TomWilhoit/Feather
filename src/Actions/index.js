export const addLocations = (locations) => ({
  type: 'ADD_LOCATIONS',
  locations
})

export const addWeather = (weatherLocals) => ({
  type: 'ADD_WEATHER',
  weatherLocals
})

