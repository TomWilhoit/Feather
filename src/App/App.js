import React, { Component } from 'react';
import { Header } from '../Header/Header';
import { Route, Switch } from "react-router-dom";
import  CardContainer  from '../CardContainer/CardContainer'
import { connect } from 'react-redux';
import { addLocations } from '../Actions'
import { addWeather } from '../Actions'
import { APIkey } from '../Utils/APIkey'
import { fetchData } from '../Utils/API'
import PropTypes from "prop-types"
import './App.scss';
import   CardDetail   from '../CardDetail/CardDetail';
export class App extends Component {
  
  componentDidMount( ){
    this.fetchBackendData();
  }

  fetchBackendData = async () => {
    const url = "http://localhost:3000/api/v1/locations"
    const response = await fetchData(url)
    await this.props.addLocations(response)
    this.fetchWeatherData(response)
  }

  fetchWeatherData = async (locations) => {
    try {
      const result = await Promise.all(locations.map( async (location) => {
        let lat = location.lat;
        let long = location.long;
        const url = `https://cors-anywhere.herokuapp.com/https://api.darksky.net/forecast/${APIkey}/${lat},${long}`
        return await fetchData(url)}))
        this.props.addWeather(result)
    } catch(error){
    } 
  }

  render() {
      return (
        <div className="App">
          <Header/>
          <Switch>
            <Route path="/" component={CardContainer}/>
            <Route path='/:id' render={({match}) => {
              const { id } = match.params
              const place = this.props.locations.find(location => location.id === id)
              if(place) {
                return <CardDetail  {...place} /> 
              }
            }} />
            />
          </Switch>
        </div>
      );
    }
  }

  App.propTypes = {
    locations: PropTypes.array,
    weather: PropTypes.array
  };

export const mapStateToProps = state => ({
  locations: state.locations,
  weather: state.weather
});

export const mapDispatchToProps = dispatch => ({
  addLocations: locations => dispatch(addLocations(locations)),
  addWeather: weatherLocals => dispatch(addWeather(weatherLocals))
})


export default connect(
  mapStateToProps,
  mapDispatchToProps
)(App);
