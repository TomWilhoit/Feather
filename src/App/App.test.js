import React from "react";
import ReactDOM from "react-dom";
import { mapStateToProps } from "./App";
import { mapDispatchToProps } from "./App";
import { shallow } from "enzyme";
import { App } from "./App";
import { addLocations, addWeather } from '../Actions/index'


describe("App", () => {
  let wrapper;
  let mockProps = {
    addWeather: jest.fn(),
    addLocations: jest.fn(),
    fetchBackendData: jest.fn(),
    locations: [
      { id: 0, lat: 25, long: 50 },
      { id: 1, lat: 30, long: 60 },
      { id: 2, lat: 35, long: 70 }
    ]
  };

  beforeEach(() => {
    wrapper = shallow(<App {...mockProps} />);
  });
  it("should match the snapshot with all data passed in correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it.skip('should fire fetchBackEndData inside of componentDidMount', () => {
    wrapper.instance().componentDidMount()
    expect(mockProps.fetchBackendData).toHaveBeenCalled()
  })

  it.skip('should fire fetchBackendData with the right URL',  () => {
    const mockUrl = 'www.reddit.com'
    wrapper.fetchBackendData()
    expect(fetchBackEndData).toHaveBeenCalledWith(mockUrl)
  })

  it.skip('should fire fetchWeather with the right URL',  () => {
    const mockUrl = 'www.reddit.com'
    wrapper.fetchWeatherData()
    expect(fetchWeatherData.toHaveBeenCalledWith(mockUrl))
  })
  


  describe("mapStateToProps", () => {
    it("should mapStateToProps", () => {
      const mockState = {
        locations: ["Chattanooga", "Memphis"],
        weather: ["sunny", "rainy"]
      };
      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(mockState);
    });
  });
  describe("mapDispatchToProps", () => {
    it("should map dispatch to props", () => {
      const mockLocations = ["Chattanooga", "Memphis"]
      const mockDispatch = jest.fn();
      const actionToDispatch = addLocations(mockLocations);
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.addLocations(mockLocations);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it("should map dispatch to props", () => {
      const mockWeather = ["sunny", "rainy"]
      const mockDispatch = jest.fn();
      const actionToDispatch = addWeather(mockWeather);
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.addWeather(mockWeather);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });
});
