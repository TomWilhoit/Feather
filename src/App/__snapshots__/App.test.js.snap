// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot with all data passed in correctly 1`] = `
<div
  className="App"
>
  <Header />
  <Switch>
    <Route
      component={
        Object {
          "$$typeof": Symbol(react.memo),
          "WrappedComponent": [Function],
          "compare": null,
          "displayName": "Connect(CardContainer)",
          "type": [Function],
          Symbol(Symbol.hasInstance): [Function],
        }
      }
      path="/"
    />
    <Route
      path="/:id"
      render={[Function]}
    />
    /&gt;
  </Switch>
</div>
`;
